README.rst

```diff
@@ -12,77 +12,90 @@ smart_open — utils for streaming large files in Python
 What?
 =====
 
-``smart_open`` is a Python 2 & Python 3 library for **efficient streaming of very large files** from/to S3, HDFS, WebHDFS, HTTP, or local (compressed) files. It's a drop-in replacement for Python's built-in ``open()``: it can do anything ``open`` can (100% compatible, falls back to native ``open`` wherever possible), plus lots of nifty extra stuff on top.
+``smart_open`` is a Python 2 & Python 3 library for **efficient streaming of very large files** from/to S3, HDFS, WebHDFS, HTTP, or local (compressed) files.
+It's a drop-in replacement for Python's built-in ``open()``: it can do anything ``open`` can (100% compatible, falls back to native ``open`` wherever possible), plus lots of nifty extra stuff on top.
 
 ``smart_open`` is well-tested, well-documented, and has a simple, Pythonic API:
 
 .. code-block:: python
 
-  >>> from smart_open import smart_open
-
+  >>> from smart_open import open
```

Comment:
-1: what's the reason for this renaming? Both confusing (shadows a built-in) and backward incompatible.

Reply:
> -1: what's the reason for this renaming?

It's not really a renaming. It's a new function. The old function still works exactly the same as before, we just don't want people to use it anymore (for reasons, see below).

> Both confusing (shadows a built-in)

I don't think it's any more confusing than, for example, io.open or any of the other similarly named functions in the [standard library](https://docs.python.org/3/search.html?q=open&check_keywords=yes&area=default). There are more than a dozen such cases, so I think the precedent for using the name `open` is strong. 

Do you recommend to do this instead:

```python
import smart_open
fin = smart_open.open(...)
```

>>> and backward incompatible.

Yes, it's backwards incompatible with the old smart_open.smart_open, but that function several big disadvantages:

- in was incompatible the built-in open function, contradicting the main mission of the project. For example, it opened things as binary by default.
- It accepted keyword arguments without checking them (via **kwargs). This made the parameters difficult to document.

In summary, the new function solves the above two problems, so it's **incompatible by design**. The old function still works the same way, so why would the incompatibility be a problem?

In summary, this new function is a **true** replacement for the built-in `open` function. It accepts the same parameters that the built-in `open` does. It also mirrors alternative entry points like `io.open` and over a dozen of other `*.open`, so I think calling it `open` it natural.

Reply:
I really hope that the users will not do `%pylab`-style star-importing everything, as it will be yet another gotcha with porting Jupyter notebooks to projects, and debugging in general.

Reply:
@vmarkovtsev I don't see the merit of your argument, for three reasons.

First, think about this: what could go wrong (TM)? The smart_open.open function behaves **identically** to the built-in open - it even falls back to the built-in open whenever possible. In the vast majority of cases, it will give **exactly the same result**.

Second, there really is no practical benefit to `from smart_open import *`, given there are only two public functions in there (the other one being s3_iter_buckets).  

Finally, star-importing everything is [generally discouraged](https://stackoverflow.com/questions/2386714/why-is-import-bad), anyway. Yeah, people still do it, but the consequences are on them.

Reply:
This is the thing, many people still do it. I am not for or against `open` btw, just writing that this will definitely cause problems in some workflows. Ours, for example.

Question: if one wants to read a gzip file verbatim, and `open` is already overwritten, they will have to use `io.open`, correct?

Reply:
Opening a gzip file in binary mode with smart_open.open will give the same results as io.open. This is one of the cases where smart_open falls back to io.open. I expect this is the use case you're referring to.

Opening a gzip file in text mode *will* give different results. smart_open will decode the file transparently. io.open will give you a UnicodeError trying to decode the bytes into meaningful text. It's unlikely your workflow does this, as the UnicodeError is not recoverable at that stage.

Reply:
As long as normal `from smart_open import smart_open` continues to work, it's probably OK (too much existing code relies on it).